src/Migration.ts

```diff
@@ -0,0 +1,305 @@
+						for (const span of node.spans) {
+							if (
+								span.type === "hyperlink" &&
+								span.data &&
+								"link_type" in span.data &&
+								span.data.link_type === LinkType.Media
+							) {
+								onAsset(span.data)
+							}
+						}
+					}
+				}
+			}
+		} else if (is.group(field)) {
+			for (const item of field) {
+				discoverAssets(item, onAsset)
+			}
+		} else if (
+			is.image(field) &&
+			field &&
+			"dimensions" in field &&
+			isFilled.image(field)
+		) {
+			onAsset(field)
+		} else if (
+			is.link(field) &&
+			field &&
+			"link_type" in field &&
+			field.link_type === LinkType.Media &&
+			isFilled.linkToMedia(field)
+		) {
+			onAsset(field)
+		}
+	}
+}
+
+/**
+ * Extracts one or more Prismic document types that match a given Prismic
+ * document type. If no matches are found, no extraction is performed and the
+ * union of all provided Prismic document types are returned.
+ *
+ * @typeParam TMigrationDocuments - Prismic migration document types from which
+ *   to extract.
+ * @typeParam TType - Type(s) to match `TMigrationDocuments` against.
+ */
+type ExtractMigrationDocumentType<
+	TMigrationDocuments extends PrismicMigrationDocument,
+	TType extends TMigrationDocuments["type"],
+> =
+	Extract<TMigrationDocuments, { type: TType }> extends never
+		? TMigrationDocuments
+		: Extract<TMigrationDocuments, { type: TType }>
+
+type CreateAssetReturnType = ImageMigrationField & {
+	image: ImageMigrationField
+	linkToMedia: LinkToMediaMigrationField
+}
+
+const SINGLE_KEY = "__SINGLE__"
+
+/**
+ * A helper that allows preparing your migration to Prismic.
+ *
+ * @typeParam TDocuments - Document types that are registered for the Prismic
+ *   repository. Query methods will automatically be typed based on this type.
+ */
+export class Migration<
+	TDocuments extends PrismicDocument = PrismicDocument,
+	TMigrationDocuments extends
+		PrismicMigrationDocument<TDocuments> = PrismicMigrationDocument<TDocuments>,
+> {
+	/**
+	 * @internal
+	 */
+	documents: {
+		document: TMigrationDocuments
+		params: PrismicMigrationDocumentParams
+	}[] = []
+	#indexedDocuments: Record<string, Record<string, TMigrationDocuments>> = {}
+
+	/**
+	 * @internal
+	 */
+	assets: Map<MigrationAsset["file"], MigrationAsset> = new Map()
+
+	createAsset(
+		asset: Asset | FilledImageFieldImage | FilledLinkToMediaField,
+	): CreateAssetReturnType
+	createAsset(
+		file: MigrationAsset["file"],
+		filename: MigrationAsset["filename"],
+		params?: {
+			notes?: string
+			credits?: string
+			alt?: string
+			tags?: string[]
+		},
+	): CreateAssetReturnType
+	createAsset(
+		fileOrAsset:
+			| MigrationAsset["file"]
+			| Asset
+			| FilledImageFieldImage
+			| FilledLinkToMediaField,
+		filename?: MigrationAsset["filename"],
+		{
+			notes,
+			credits,
+			alt,
+			tags,
+		}: {
+			notes?: string
+			credits?: string
+			alt?: string
+			tags?: string[]
+		} = {},
+	): CreateAssetReturnType {
+		let asset: MigrationAsset
```

Comment:
I thought about it a bit at the beginning and decided against it as it does not really match our style.

It seems like it could help bit but I'm not sure it would save a lot of code as bulk of the complexity is related to regular fields, detecting "migration" types is quite straightforward and simple next to that.

However, since you mentioned it multiple time now I feel like you have a feeling about using classes, so maybe we can discuss it further to see how we could align on that?